contracts/operators/CollectionBatchBuyOperator.sol

```diff
@@ -147,38 +164,36 @@ contract CollectionBatchBuyOperator is IOperator {
             }
 
             {
-                uint96 ethUsed;
+                uint96 receivedTokensValue = 0;
                 for (uint256 j; j < call.tokensToBuy.length; ++j) {
                     uint256 tokenId = call.tokensToBuy[j].tokenId;
 
-                    // Check whether the NFT was successfully bought.
                     if (op.nftContract.safeOwnerOf(tokenId) == receiver) {
-                        ethUsed += call.tokensToBuy[j].price;
-                        ++tokensBought;
+                        purchasedTokenIds[numTokensBought++] = tokenId;
+                        receivedTokensValue += call.tokensToBuy[j].price;
 
-                        // Add the token to the list of tokens to finalize.
-                        tokenIds[tokensBought - 1] = tokenId;
+                        if (!ex.isReceivedDirectly) {
+                            // Transfer the NFT to the Party.
+                            op.nftContract.safeTransferFrom(address(this), msg.sender, tokenId);
+                        }
                     }
                 }
-
-                // Check ETH spent for call is what was expected.
-                uint256 actualEthUsed = balanceBefore - address(this).balance;
-                if (ethUsed != actualEthUsed) {
-                    revert EthUsedForFailedBuyError(ethUsed, actualEthUsed);
+                uint256 callUsage = beforeCallBalance - address(this).balance;
+                if (callUsage > receivedTokensValue) {
+                    revert EthUsedForFailedBuyError(receivedTokensValue, callUsage);
                 }
-
-                totalEthUsed += ethUsed;
             }
         }
 
         // This is to prevent this crowdfund from finalizing a loss if nothing
-        // was attempted to be bought (ie. `tokenIds` is empty) or all NFTs were
+        // was attempted to be bought (ie. `purchasedTokenIds` is empty) or all NFTs were
         // bought for free.
+        uint256 totalEthUsed = beforeEthBalance - address(this).balance;
```

Comment:
Nit, but we could move `uint256 unusedEth = address(this).balance;` above this line and update it to:
```suggestion
        uint256 totalEthUsed = beforeEthBalance - unusedEth;
```

Reply:
yup. Much better